scripts/ci/dag.py

```diff
@@ -0,0 +1,169 @@
+from __future__ import annotations
+
+import time
+from concurrent.futures import ThreadPoolExecutor
+from multiprocessing import Event, cpu_count
+from multiprocessing.synchronize import Event as EventClass
+from queue import Empty, Queue
+from typing import Callable, Generic, Hashable, TypeVar
+
+_T = TypeVar("_T", bound=Hashable)
+
+
+class _Node(Generic[_T]):
+    def __init__(self, value: _T):
+        self.value = value
+        self.counter: int = 0
+        self.dependents: list[_Node[_T]] = []
+
+
+class DAG(Generic[_T]):
+    def __init__(self, dependency_graph: dict[_T, list[_T]]):
+        """
+        Construct a directed acyclic graph from an adjacency list.
+
+        The `dependency_graph` _must not_ contain any cycles.
+        """
+
+        self._nodes: dict[_T, _Node[_T]] = {}
+        self._queue: list[_T] = []
+        self._num_finished: int = 0
+
+        for node, deps in dependency_graph.items():
+            new_node = self._get_or_insert(node)
```

Comment:
let's be consistent with the naming, and not use the same name for two things

```suggestion
        for value, deps in dependency_graph.items():
            new_node = self._get_or_insert(value)
```